### PR TITLE
add support for a custom error check for retries

### PR DESCRIPTION
**Pull request by pmenglund**
support providing a custom error check function for retries

### CODE REVIEW

**Comment by pmenglund** [comment]
Current dependencies on/for this PR:
* master
  * **PR #75** <a href="https://app.graphite.dev/github/pr/rockset/rockset-go-client/75" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/rockset/rockset-go-client/75?utm_source=stack-comment).

File: retry.go [hunk 1]
```diff
@@ -118,8 +135,8 @@ func (r ExponentialRetry) RetryWithCheck(ctx context.Context, checkFn RetryCheck
 		waitInterval = r.WaitInterval
 	}
 	jitterFraction := .05
-	if r.jitterFraction != 0 {
-		jitterFraction = r.jitterFraction
+	if r.JitterFraction != 0 {
+		jitterFraction = r.JitterFraction
```
Comment (jadunham1) [inline retry.go:139]: Github won't let me comment on line #149 but I'm guessing it should be `call duration`

**Review by jadunham1 — COMMENTED** [review]
Is the expected behavior to retry until success?  For some reason I can't find any `max attempts` type of variable.

**Review by morganda — APPROVED** [review]
lgtm

til - structs need whatever constructs an instance to assign the default values which is why using a method to construct a new object is so useful so that you can make sure new instances have sane defaults

File: retry.go (hunk 1 above)
Comment (pmenglund) [reply retry.go:139]: if you want to set a time limit, you do it through the context, but a max retries would make sense

```
ctx := context.WithTimeout(parentCtx, 5*time.Second)
err = rc.Retry(ctx, func() error{ ... })
```

**Comment by pmenglund** [comment]
if you want to set a time limit, you do it through the context, but a max retries would make sense

```
ctx := context.WithTimeout(parentCtx, 5*time.Second)
err = rc.Retry(ctx, func() error{ ... })
```

**Comment by jadunham1** [comment]
> if you want to set a time limit, you do it through the context, but a max retries would make sense
> 
> ```
> ctx := context.WithTimeout(parentCtx, 5*time.Second)
> err = rc.Retry(ctx, func() error{ ... })
> ```

Makes sense, I wouldn't suspect this is the way I'd use a retryable is via the context, but perhaps those more fluent in go would assume this.

**Pull request merged**